### ISSUE TITLE
API to check R package version

### ISSUE THREAD

**Issue by steffilazerte** [MotusDev/Motus-TO-DO#914]
This issue was originally raised in #760, but for clarity I've moved the relevant sections here so we can work on it separately.

**TL;DR**
The idea here is that the R package should send it's version to the API, and the API will check if it's

a) up to date (no message; OR status message)
b) out of date but working (update message/warning)
c) out of date and not under warranty (update warning/error)

R will send the version as 0.4.1.9999 (etc.) unless the API needs a different format for comparison (i.e. R recognizes 0.4.12 as greater than 0.4.9, but maybe the API doesn't)?

- [x] @steffilazerte Add the version number to motus authentication API call (will send as `pkgVersion = "0.4.12"` etc.)
- [x] @denislepage or @cjardine-bsc Start tracking the minimum and current R package versions in the database
- [ ] @pmorrill (@steffilazerte to confirm messages/warnings/errors) Validate package version in API authentication call according to above options

Sound good?

---
**Excerpts from original issue**

From @steffilazerte ([original comment](https://github.com/MotusDev/Motus-TO-DO/issues/760#issuecomment-680266513)):

>  ... Have the server send the motus package a minimum package version (perhaps as part of the auth?) which the R package checks against. If the current version is less than the server version, the download errors and prompts the user to update their package. However, this will only work on future versions of the motus package. Here the error is produced by R
        Alternatively, the motus package could send it's version to the server, and if the server receives a low version or no version it returns an error that prompts the user to update. Here the error is produced by the server and R just shows it to the user.

From @denislepage ([original comment](https://github.com/MotusDev/Motus-TO-DO/issues/760#issuecomment-680290574)):
> I think that adding the R package version to the API call is a good idea. We could keep track of the minimum required package version supported and the latest version available.

> If the package version is below the minimum required, we could force the user to update the package and return an error message and no data. Perhaps we need to have a way for the user to bypass the error (with no guarantee it will work).  E.g. “Your motus package version (xx) is no longer supported. Please update your package to continue accessing Motus data. To ignore this error, you can add the ignoreVersion=true parameter, but this may lead to unexpected results.” (if we decide to ignore, we could change the error to a warning).

> If the package version is below the latest version, but higher than the minimum required, we could send a warning to be displayed (e.g. a newer motus package is available, please update your package as soon as possible)

> We certainly don’t want to display this message with every single API call, but ideally, it would be once for every function call made by the user. Perhaps only the authentication to start is enough. Worst case if needed, we could then terminate all tokens and force people to reissue an authentication call, or refuse/warn all calls made that do not supply the version number, which will force/encourage all the legacy users to upgrade.

> Can you easily integrate the version number as part of your deployment process, and add that number to the authentication call? We’ll eventually deal with the handling on the API side. Sounds simple enough from your end. The one thing we’ll need to do is remember to track the latest and minimum version numbers on the database side.

> Final point, is the version number format easily comparable right now (e.g. decimal 1.2) or do we need to device some rules for comparing 2 versions (e.g. 1.1.10 may appear earlier than 1.1.9 if compared as strings).

> So:
> - Add the version number to the Motus package constant if not already there and confirm the version number format (Steffi)
> - Add the version number to any submissions to the authenticate API call (Steffi)
> - Start tracking the minimum and current R package versions in the database. I feel like we already have a place where this could fit, but if not, we should create a generalized table with key/values for global variables. (Catherine/me)
> - Use a validation rule in the API authentication call so we can check if 1) the version is lower than the minimum (returns an error and no data) and 2) the version is lower than the current package version (returns data and a warning to be displayed on the R package) (Paul, with Steffi to confirm on the warning display)

**Comment by denislepage** [comment]
We'll need to build a couple lines of custom code to compare version parts, but that should be simple enough.

For safety, we can probably even support an even number of levels.

4.19.12 would be larger than 4.19 for instance (assuming the missing parts are 0)

Hopefully the parts are always numeric, yes?

**Comment by steffilazerte** [comment]
Options are always 3 or 4 parts:

4.19.0
4.19.0.9000 (dev)

Always numeric :)

**Comment by denislepage** [comment]
@pmorrill , I will add this to the core bscbase methods below. Making this into a comparator potentially allows us to sort a list of versions, etc.

```

package org.bsc.comparators;

public class CompareVersionNumbers implements java.util.Comparator<String> {

	private String delim = "\\.";  
	public int compare(String s1, String s2) {
		
		String[] v1 = s1.split(delim);
		String[] v2 = s2.split(delim);
		
		for (int i = 0; i < v1.length && i < v2.length; i++) {
			int c =Integer.valueOf(v1[i]).compareTo(Integer.valueOf(v2[i])); 
			if (c != 0) return c;  
		}
		// if we make it here, the initial parts are the same
		if (v1.length > v2.length) return 1; 
		if (v1.length < v2.length) return -1;
		else return 0;
	}

	public CompareVersionNumbers() {
		super();
	}
	public CompareVersionNumbers(String delim) {
		super();
		this.delim = delim;
	}
	
	public static void main(String[] s) {

		CompareVersionNumbers cvn = new CompareVersionNumbers();
		System.out.println(cvn.compare("4.5.10", "4.5"));

	}
	
}
```

```
	CompareVersionNumbers cvn = new CompareVersionNumbers();
	System.out.println(cvn.compare("4.5.10", "4.5"));

```
Returns 1 when s1 > s2, -1 when s1 < s2 and 0 when they are equal. For this purpose, 4.5.0001 should be the same as 4.5.1, and 4.5.8000 will be greater than  4.5.900. 

4.4.100 will also obviously be smaller than 4.5.0.

**Comment by denislepage** [comment]
pushed it to git

**Comment by pmorrill** [comment]
I see the code in git. I assume Steffi will add the R version to auth call, but how do we track the status of different versions? Presumably, a db table. Once I know that, I can add the code to make a comparison.

**Comment by pmorrill** [comment]
@steffilazerte would it be useful if I added a simple api entrypoint that allows you to SET the current version into our database? Could you incorporate that into the build process when you deploy a new version of the client? 

That would be allow you to keep the CURRENT version up to date, without requiring emails to Catherine / Denis to coordinate that value.

If this would be useful to you, we will stick-handle a permission role attached to your auth. profile.

**Comment by cjardine-bsc** [comment]
I'll set you up a table. What is the current version and the lowest version number we want to support?

**Comment by steffilazerte** [comment]
@pmorrill I think that could work

How does this work with the different branches? Presumably each would have their own minimum and current version?

**Comment by denislepage** [comment]
I just remembered that we already keep track of the current version in the adm_platform_api, though we are not maintaining it nearly often as we should. There’s a version and a releaseTime (date) field that we should try to keep current.

The current version is available from the github repo (4.0.2). I’d love if this could be included into the build process.

Not sure if adding a field there for the minimum version supported would mess up the CANARIE API, but that’s an option to consider. We could have a view to exclude certain fields for the CANARIE API needs. Another benefit of using the same table is that we could potentially also automate the update for the other platform entries the same way. I just updated the version numbers, but most of them were at least 2 years behind.

Our table current has 5 entries.

id

baseUrl

name

version

releaseTime

1

platform

Motus Wildlife Tracking System

9.21.01.14

2021-01-14

2

r-packages/service

Motus R Packages

4.0.2

2020-10-27

3

client-api/service

Motus Client Data API

9.21.01.14

2021-01-14

4

server-api/service

Motus Server Data API

9.21.01.14

2021-01-14

5

sensorgnome/services

Motus SensorGnome Software

2021-02-10

2021-02-10

**Comment by steffilazerte** [comment]
Okay, for testing, this goes on sanbox then? Or does it matter?

**Comment by pmorrill** [comment]
@denislepage your CompareVersionNumbers class does not appear in the current bsc-base library in motus/shared-lib. Can you make sure that bsc-base version is current in git?

@Hussell may be able to check it easily as well.

**Comment by denislepage** [comment]
I thought it was. It’s under org.bsc.comparators.CompareVersionNumbers

I can see it in the git repo on the server.

**Comment by denislepage** [comment]
Oh. I see what you were saying. Perhaps the compiled bscbase version in Motus hadn’t been updated yet? I had pushed my changes last week, but I am not sure if Jeremy only updates the bcsbase jar as needed.

I have recompiled the jar and pushed it to the motus sandbox branch. That should do it.

**Comment by pmorrill** [comment]
I have updated the sandbox api as follows:

- our stored pgkVersion (taken from adm_platform_api) will be returned by the /api_info function, and also by the /authenticate_user function; The returned field is called 'currentPkgVersion';

- the /authenticate_user function now expects to get a parameter 'pkgVersion' from the R-client. If this is null or less than our currentPkgVersion, the response package will include a warning and a suggestion that the user update the R-client software (we can fine-tune the wording in the adm_errors system);

- there is a new function /update_pkg_version that expects a token and a pgkVersion parameter. If the pgkVersion field is greater than the current stored version, the latter will be updated, and a success will be returned along with the new version as stored in the database;

The /update_pkg_version function currently requires that the user have admin privileges. Do we need set up a role for @steffilazerte or is she an admin?

These changes will be available after the next sandbox deploy.

**Comment by denislepage** [comment]
I’d go with a new role for updating the version.

**Comment by steffilazerte** [comment]
Okay, I've added partial functionality. In sandbox, the motus package will warn users if they're package version is too low (i.e. if the server returns warning. But I haven't looked into updating the version.

Do we have a version below which we error?

**Comment by denislepage** [comment]
Not yet, no. Any of the older releases that we might want to deprecated won’t show the error anyway, so this would only apply to new changes we make moving forward that breaks the package functionalities as they exist now. The other thing is that older packages won’t be sending their version number.

There’s potentially a point where versionless requests from the older R packages will be rejected and presented with an error message, but that would also potentially break other possible applications of the API outside of R, so we’ll need to weigh that carefully.

For now, warning it is.

**Comment by steffilazerte** [comment]
Good point!

Perhaps we should put in this functionality now (even if it would currently only work on bogus versions), so I can add the corresponding functionality to the R package. This way when *this* version of the R package is so old we want to force an update, we'll have put a graceful way of doing so in the package, without having to just error on all NULL package versions...?

**Comment by steffilazerte** [comment]
@pmorrill:
> - there is a new function /update_pkg_version that expects a token and a pgkVersion parameter. If the pgkVersion field is greater than the current stored version, the latter will be updated, and a success will be returned along with the new version as stored in the database;
>
> The /update_pkg_version function currently requires that the user have admin privileges. Do we need set up a role for @steffilazerte or is she an admin?

Was this ever resolved? I tried to access https://sandbox.motus.org/api/sgdata/update_pkg_version with my steffilazerte user name but was told I don't have access. 

By the way, would it be possible to get permission to also *down-grade* the package requirements? I'm nervous that I'll accidentally update the package version on the master branch to that required by sandbox which will result in annoying (and impossible) update messages for everyone. It'd be nice if I could immediately fix those mistakes without having to track down someone with server access!

**Comment by pmorrill** [comment]
The api entrypoint currently only allows you to update - it will return an error if you try to set the version lower than it already is. It's pretty easy to change. Thoughts @denislepage  ?

**Comment by denislepage** [comment]
Giving Steffi the ability to downgrade the version requirement makes sense.

At this point, we’re only using this as a signal for people to upgrade, and not a strict requirement, so the only impact of the version number is the message displayed for users on the R side.

**Comment by pmorrill** [comment]
OK - done. Just checked in to sandbox.

Note that none of this pgk-version-update functionality is in beta9 (or live) yet. If the comparator class is in the bsc-base lib we use there, I can try to move this in?

**Comment by denislepage** [comment]
I’m presuming that Jeremy keeps the bscbase classes up to date in beta9.

As discussed, I’d be tempted to push a new beta10 branch to the beta server so we can test it. @Hussell, could you create a new beta10 branch early next week. That will allow me to test at least the processing of dual model files and get through the backlog of those files. 

We can revert back to beta9 after that if necessary, which I expect will be the case.

We could potentially also put Lucas's dashboard stuff in beta11 and separate the accumulated changes from the new stuff.
